<launch>

   <node name="topic_publisher" pkg="indoor_navigation" type="test.py" output="screen">
   </node>

    <!--<node pkg="ytnode" type="ytnode" name="ytnode" output="screen" />
    <param name="p1" value="1"/>
    <param name="p2" value="2"/>
   </node> -->
   
   <include file="$(find velodyne_pointcloud)/launch/VLP16_points.launch">
   </include>
   
   <node pkg="my_serial_node" type="my_serial_node" name="my_serial_node" output="screen" >
    <param name="num" value="14"/>
   </node>

   <node name="gps_conv" pkg="gps_common" type="utm_odometry_node">
    <param name="rot_covariance" value="99999" />
    <param name="frame_id" value="base_link" />
   </node>

   <node pkg="control_node" type="control_node" name="control_node" output="screen" >
    <param name="cp0"     value="1e-3"/>
    <param name="cp7"     value="1e-3"/>
    <param name="cp8"     value="1e-9"/>
    <param name="cp14"     value="1e6"/>
    <param name="cp21"     value="1e6"/>
    <param name="cp28"     value="1e6"/>
    <param name="cp35"     value="1e-9"/>
    <param name="ct0"     value="1e-9"/>
    <param name="ct7"     value="1e-3"/>
    <param name="ct8"     value="1e-9"/>
    <param name="ct14"     value="1e6"/>
    <param name="ct21"     value="1e6"/>
    <param name="ct28"     value="1e6"/>
    <param name="ct35"     value="1e-9"/>
    <param name="cpv0"     value="1e-3"/>
    <param name="cpv8"     value="0"/>
    <param name="cpv35"     value="1e3"/>
    <param name="ctv0"     value="1e-3"/>
    <param name="ctv8"     value="0"/>
    <param name="ctv35"     value="1e3"/>
   </node>

   <node name="imu_read_node" pkg="miiboo_imu"  type="imu_read" output="screen">
    <param name="c0"     value="1e6"/>
    <param name="c4"     value="1e6"/>
    <param name="c8"     value="1e-6"/>
    <param name="avc0"     value="1e6"/>
    <param name="avc4"     value="1e6"/>
    <param name="avc8"     value="1e-6"/>
   </node>
   
    <node name="map_server" pkg="map_server" type="map_server" args="$(find indoor_navigation)/maps/blank_map.yaml"/>

    <param name="/use_sim_time" value="false"/>

   
   
   <node pkg="tf" type="static_transform_publisher" name="base_link_to_laser" 
    args="0.52 0.0 -0.175 0.0 0.0 0.0 /base_link /velodyne 40" />

   <node pkg="tf" type="static_transform_publisher" name="base_footprint_to_base_link" 
    args="0.0 0.0 0.0 0.0 0.0 0.0 /base_footprint /base_link 40" />

   <node pkg="tf" type="static_transform_publisher" name="base_footprint_to_imu" 
    args="0.52 0.0 -0.135 0.0 0.0 0.0 /base_footprint /imu_link 40" />

   <node pkg="tf" type="static_transform_publisher" name="map_to_odom" 
    args="0.0 0.0 0.0 0.0 0.0 0.0 /map /odom 40" />
   

   <node pkg="robot_pose_ekf" type="robot_pose_ekf" name="robot_pose_ekf">
    <param name="output_frame" value="odom"/>
    <param name="base_footprint_frame" value="base_footprint"/>
    <param name="freq" value="30.0"/>
    <param name="sensor_timeout" value="1.0"/>
    <param name="dey" value="0.5"/>
    <param name="dex" value="0.5"/>
    <param name="detheta" value="1.2"/>
    <param name="odom_used" value="true"/>
    <param name="imu_used" value="true"/>
    <param name="vo_used" value="false"/>
    <param name="gps_used" value="true"/>
    <param name="debug" value="false"/>
    <param name="self_diagnose" value="false"/>
   </node>

   <node pkg="rviz" type="rviz" name="rviz" 
    args="-d $(find control_node)/demo_gmapping.rviz"/>

   <!--<node pkg="robot_pose_ekf" type="robot_pose_ekf" name="robot_pose_ekf">
    <param name="output_frame" value="odom"/>
    <param name="freq" value="10.0"/>
    <param name="sensor_timeout" value="1.0"/>
    <param name="odom_used" value="true"/>
    <param name="imu_used" value="true"/>
    <param name="vo_used" value="false"/>
    <param name="debug" value="false"/>
    <param name="self_diagnose" value="false"/>
  </node>-->
  

  <arg name="use_map_topic" default="false"/>
  <arg name="scan_topic" default="scan"/>


  <node pkg="amcl" type="amcl" name="amcl">
    <param name="use_map_topic" value="$(arg use_map_topic)"/>
    <!-- Publish scans from best pose at a max of 10 Hz -->
    <param name="odom_model_type" value="diff"/>
    <param name="odom_alpha5" value="0.1"/>
    <param name="gui_publish_rate" value="10.0"/>
    <param name="laser_max_beams" value="60"/>
    <param name="laser_max_range" value="12.0"/>
    <param name="min_particles" value="500"/>
    <param name="max_particles" value="2000"/>
    <param name="kld_err" value="0.05"/>
    <param name="kld_z" value="0.99"/>
    <param name="odom_alpha1" value="0.2"/>
    <param name="odom_alpha2" value="0.2"/>
    <!-- translation std dev, m -->
    <param name="odom_alpha3" value="0.2"/>
    <param name="odom_alpha4" value="0.2"/>
    <param name="laser_z_hit" value="0.5"/>
    <param name="laser_z_short" value="0.05"/>
    <param name="laser_z_max" value="0.05"/>
    <param name="laser_z_rand" value="0.5"/>
    <param name="laser_sigma_hit" value="0.2"/>
    <param name="laser_lambda_short" value="0.1"/>
    <param name="laser_model_type" value="likelihood_field"/>
    <!-- <param name="laser_model_type" value="beam"/> -->
    <param name="laser_likelihood_max_dist" value="2.0"/>
    <param name="update_min_d" value="0.25"/>
    <param name="update_min_a" value="0.2"/>
    <param name="odom_frame_id" value="odom"/>
    <param name="resample_interval" value="1"/>
    <!-- Increase tolerance because the computer can get quite busy -->
    <param name="transform_tolerance" value="1.0"/>
    <param name="recovery_alpha_slow" value="0.0"/>
    <param name="recovery_alpha_fast" value="0.0"/>
    <remap from="scan" to="$(arg scan_topic)"/>
  </node>

  <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping" output="screen">
    <param name="map_udpate_interval" value="1.0"/>
    <param name="maxUrange" value="10.0"/>
    <param name="sigma" value="0.1"/>
    <param name="kernelSize" value="1"/>
    <param name="lstep" value="0.15"/>
    <param name="astep" value="0.15"/>
    <param name="iterations" value="1"/>
    <param name="lsigma" value="0.1"/>
    <param name="ogain" value="3.0"/>
    <param name="lskip" value="1"/>
    <param name="srr" value="0.1"/>
    <param name="srt" value="0.2"/>
    <param name="str" value="0.1"/>
    <param name="stt" value="0.2"/>
    <param name="linearUpdate" value="1.0"/>
    <param name="angularUpdate" value="0.5"/>
    <param name="temporalUpdate" value="0.4"/>
    <param name="resampleThreshold" value="0.5"/>
    <param name="particles" value="10"/>
    <param name="xmin" value="-5.0"/>
    <param name="ymin" value="-5.0"/>
    <param name="xmax" value="5.0"/>
    <param name="ymax" value="5.0"/>
    <param name="delta" value="0.032"/>
    <param name="llsamplerange" value="0.01"/>
    <param name="llsamplestep" value="0.05"/>
    <param name="lasamplerange" value="0.05"/>
    <param name="lasamplestep" value="0.05"/>
  </node>

  <node pkg="move_base" type="move_base" respawn="false" name="move_base" clear_params="true">
    <rosparam file="$(find indoor_navigation)/config/config_default/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find indoor_navigation)/config/config_default/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find indoor_navigation)/config/config_default/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find indoor_navigation)/config/config_default/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find indoor_navigation)/config/config_default/base_local_planner_params.yaml" command="load" />
  </node>

  <node pkg="hit_nav" type="move_base_square_hit" name="nav_move_base" output="screen">
  </node>


   <!--<node name="talker" pkg="indoor_navigation" type="start.py" output="screen">
   </node>
   
   <node name="ros2jsonguiji" pkg="indoor_navigation" type="startguiji.py" output="screen">
   </node>
   
   <node name="ros2jsonmap" pkg="indoor_navigation" type="startmap.py" output="screen">
   </node>

   <node name="ros2json" pkg="indoor_navigation" type="start2.py" output="screen">
   </node>

   <node name="mrobot_teleop" pkg="mrobot_teleop" type="client.py" output="screen">
   </node>


   <node name="listener" pkg="indoor_navigation" type="lis.py" output="screen">
   </node>-->

</launch>
